onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2843,6 +2843,70 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
         }
       });
 
+  static const char* Trilu_ver1_doc = R"DOC(
+	  Returns the upper or lower triangular part of a 2-D matrix, or batches of 2-D matrices depending on whether attribute upper is true or not.
+    Tri takes one input tensor of shape [*, N, M], where * is zero or more batch dimensions. The upper/lower triangular part consists
+    of the elements on or above/below the given diagonal (k).
+	  If k = 0, this op retains elements on or above/below the the main diagonal. If k > 0, this op
+	  retains elements on or above/below k diagonals over the main diagonal. If k < 0, this op retains elements
+	  on or above/below as many diagonals below the main diagonal. All other elements in the matrix are set to zero.
+	  )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Trilu)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(Trilu_ver1_doc)
+      .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+      .Attr("upper",
+            "Boolean whether upper or lower part of matrix retains it's elements. Default value of upper is true.",
+            AttributeProto::INT,
+            static_cast<int64_t>(1))
+      .Input(
+          0,
+          "X",
+          "Input tensor of rank 2 or higher.",
+          "T")
+      .Input(
+          1,
+          "k",
+          "A 0-D tensor containing a single value corresponding to the number diagonals above or the main diagonal to exclude or include."
+          "Default value is 0 if it's not specified.",
+          "tensor(int64)",
+          OpSchema::Optional)
+      .Output(
+          0,
+          "Y",
+          "Output tensor of the same type and shape as the input tensor.",
+          "T")
+      .TypeConstraint(
+          "T",
+          {"tensor(float16)",
+           "tensor(float)",
+           "tensor(double)",
+           "tensor(bfloat16)",
+           "tensor(uint8)",
+           "tensor(uint16)",
+           "tensor(uint32)",
+           "tensor(uint64)",
+           "tensor(int8)",
+           "tensor(int16)",
+           "tensor(int32)",
+           "tensor(int64)",
+           "tensor(bool)"},
+          "Constrain input and output types to int and float tensors.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        using namespace ONNX_NAMESPACE;
+        if (hasInputShape(ctx, 0)) {
+          const TensorShapeProto& input_shape =
+              ctx.getInputType(0)->tensor_type().shape();
+          const int rank = static_cast<int>(input_shape.dim_size());
+          if (rank < 2) {
+            fail_shape_inference("Input rank must be >= 2.")
+          }
+          propagateShapeAndTypeFromFirstInput(ctx);
```

Comment:
sorry, missed this earlier: the call to propagateShapeAndTypeFromFirstInput should always happen independent of hasInputShape. It's possible the shape is not known at this time, but you do still want the element type to be set on the output. You can break up the parts if you like:

  propagateElemTypeFromInputToOutput(ctx, 0, 0);
  propagateShapeFromInputToOutput(ctx, 0, 0); // do this inside hasInputShape